### PR TITLE
propose Callback

### PR DESCRIPTION
**Pull request by rwalker-apple**
#### Problem
 Facilities for registering so-called "callbacks" are many and varied in CHIP

 #### Summary of Changes
 Propose yet another ;) .

 Seriously, though, Callbacks have:

 * caller-allocated list management and callee-registration fields that have clear "take ownership" semantics.
 * default, but configurable type signature for the callback function itself
 * auto-cancellation on destruction

 Callback assumptions/restrictions:
 * a single-threaded environment where a caller and callee live
 * callbacks can live on only one queue at a time, i.e. can be registered with only one subsystem at a time.

 Included with the tests are a couple of example use cases.  Callback should be able to support runtime registration of callbacks with the ZCL, timers, network events, etc.

### CODE REVIEW

**Review by woody-apple — APPROVED** [review]

**Review by BroderickCarlin — APPROVED** [review]

**Review by jelderton — APPROVED** [review]

**Review by andy31415 — APPROVED** [review]
Approving as a starting point, however I find the code hard to follow:
   - unit tests are using internal structures and functions
   - header comment does not provide examples with the reasoning of 'this was hard to document previously' I  find that worrysome: if it cannot be easily explained/understood, maybe some refactoring is in order.
